### PR TITLE
Integrate sentry for simulator

### LINKED ISSUE #4962: Integrate sentry for Simulator
Push errors to sentry for better tracking.

### PR DESCRIPTION
**Pull request by salmoneatenbybear**
Fixes #4962

#### Describe the changes you have made in this PR -

- Integrated Sentry into the CircuitVerse simulator to enhance error tracking capabilities.
- Added try-catch blocks in key areas to handle and report errors more effectively.
- Implemented error handling in specific modules, including `app.js`, `canvasApi.js`, `circuit.js`, `simulationArea.js`, and `testbench.js`.

image

- fixed simulator zoom slider UI that was broken after bootstrap migration from #5135

image

## Release Notes

- **New Features**
  - Integrated Sentry for enhanced error tracking across multiple components and functionalities.
  - Updated the zoom slider layout for improved visual consistency and accessibility.

- **Bug Fixes**
  - Enhanced error handling in various methods to prevent application crashes and log exceptions.

- **Documentation**
  - Corrected comments for clarity in the `simulationArea.js`.

- **Style**
  - Adjusted CSS for zoom slider buttons to improve alignment and appearance.

These updates aim to improve user experience through better error management and a more polished interface.

### CODE REVIEW

**Comment by salmoneatenbybear** [comment]
@tachyons pls review if you find time

Thankyou !

File: app/views/simulator/edit.html.erb [hunk 1]
```diff
@@ -152,12 +155,12 @@
...
         </button> -->
-        <div class="zoom-slider">
-          <button class="zoom-slider-decrement" id="decrement">-</button>
-          <input type="range" class="form-range" min="0" max="45" step="1" id="customRange1">
-          <span id='slider_value'></span>
-          <button class="zoom-slider-increment" id="increment">+</button>
-        </div>
+        <div class="zoom-slider d-flex align-items-center">
+    <button class="zoom-slider-decrement btn btn-outline-secondary" id="decrement">-</button>
+    <input type="range" class="form-range mx-2" min="0" max="45" step="1" id="customRange1">
```
[the remark this answers is not in the document]
Comment (salmoneatenbybear) [reply app/views/simulator/edit.html.erb:160]: DONE

File: simulator/src/app.js [hunk 2]
```diff
@@ -1,6 +1,15 @@
+import * as Sentry from "@sentry/browser";
 import { setup } from './setup';
 import Array from './arrayHelpers';
 
+// Initialize Sentry for error tracking
+Sentry.init({
+    dsn: "https://a9660d5d8964d5b348b07b77691d61a7@o4507251968376832.ingest.de.sentry.io/4507251969949776",
+    maxBreadcrumbs: 50,
+    debug: true,
+  });
```
[the remark this answers is not in the document]
Comment (salmoneatenbybear) [reply simulator/src/app.js:10]: DONE

File: simulator/src/app.js [hunk 3]
```diff
@@ -1,6 +1,15 @@
...
+
```
[the remark this answers is not in the document]
Comment (salmoneatenbybear) [reply simulator/src/app.js:11]: DONE

File: simulator/src/canvasApi.js [hunk 4]
```diff
@@ -7,148 +7,133 @@ import { colors } from './themer/themer';
...
+                        if (obj.absX() < simulationArea.minWidth) { simulationArea.minWidth = obj.absX(); }
+                        if (obj.absX() > simulationArea.maxWidth) { simulationArea.maxWidth = obj.absX(); }
+                    }
                 }
             }
         }
+        simulationArea.objectList = updateOrder;
+    } catch (error) {
+        Sentry.captureException(error);
+        console.error("Error captured in Sentry:", error);
```
[the remark this answers is not in the document]
Comment (salmoneatenbybear) [reply simulator/src/canvasApi.js:44]: The console.error statement is intentional to support immediate debugging during development. While errors are sent to Sentry, having them in the console helps developers troubleshoot in real-time.

File: simulator/src/canvasApi.js [hunk 5]
```diff
@@ -7,148 +7,133 @@
...
-        } else { // mouse location
-            // eslint-disable-next-line no-lonely-if
-            if (method === 1) {
-                xx = simulationArea.mouseX;
-                yy = simulationArea.mouseY;
-            } else if (method === 2) {
-                xx = (width / 2 - globalScope.ox) / globalScope.scale;
-                yy = (height / 2 - globalScope.oy) / globalScope.scale;
+    try {
+        if (method === 3) {
```
[the remark this answers is not in the document]
Comment (salmoneatenbybear) [reply simulator/src/canvasApi.js:50]: done

File: simulator/src/canvasApi.js [hunk 6]
```diff
@@ -7,148 +7,133 @@
...
+            xx = (width / 2 - globalScope.ox) / globalScope.scale;
```
[the remark this answers is not in the document]
Comment (salmoneatenbybear) [reply simulator/src/canvasApi.js:51]: done

File: simulator/src/canvasApi.js [hunk 7]
```diff
@@ -7,148 +7,133 @@
...
+            yy = (height / 2 - globalScope.oy) / globalScope.scale;
+        } else if (xx === undefined || yy === undefined || xx === 'zoomButton' || yy === 'zoomButton') {
+            if (simulationArea.lastSelected && simulationArea.lastSelected.objectType !== 'Wire') { // selected object
```
[the remark this answers is not in the document]
Comment (salmoneatenbybear) [reply simulator/src/canvasApi.js:54]: done

File: simulator/src/canvasApi.js [hunk 8]
```diff
@@ -7,148 +7,133 @@
...
+                xx = simulationArea.lastSelected.x;
```
[the remark this answers is not in the document]
Comment (salmoneatenbybear) [reply simulator/src/canvasApi.js:55]: done

File: simulator/src/canvasApi.js [hunk 9]
```diff
@@ -7,148 +7,133 @@
...
+                yy = simulationArea.lastSelected.y;
+            } else {
```
[the remark this answers is not in the document]
Comment (salmoneatenbybear) [reply simulator/src/canvasApi.js:57]: done

File: simulator/src/canvasApi.js [hunk 10]
```diff
@@ -7,148 +7,133 @@
...
+                if (method === 1) {
```
[the remark this answers is not in the document]
Comment (salmoneatenbybear) [reply simulator/src/canvasApi.js:58]: done

File: package.json [hunk 11]
```diff
@@ -40,6 +40,8 @@
   "dependencies": {
     "@fortawesome/fontawesome-free": "^6.6.0",
     "@hotwired/stimulus": "^3.2.2",
+    "@popperjs/core": "^2.11.8",
+    "@sentry/browser": "^8.37.1",
```
[the remark this answers is not in the document]
Comment (salmoneatenbybear) [reply package.json:44]: As of November 9, 2024, the latest stable version of @sentry/browser is 8.36.0, released on November 7, 2024. I have updated to the same. 

source : https://www.npmjs.com/package/%40sentry/browser

File: simulator/src/simulationArea.js [hunk 12]
```diff
@@ -71,38 +72,59 @@ const simulationArea = {
...
-            simulationArea.clickCount = 0;
-        }, 600);
+        try {
+            ckickTimer = setTimeout(() => {
+                simulationArea.clickCount = 0;
+            }, 600);
+        } catch (error) {
+            Sentry.captureException(error);
+            console.error("Error captured in Sentry:", error);
+        }
```
[the remark this answers is not in the document]
Comment (salmoneatenbybear) [reply simulator/src/simulationArea.js:82]: done

File: simulator/src/simulationArea.js [hunk 13]
```diff
@@ -71,38 +72,59 @@
...
+            if (!this.context) {
+                return;
+            }
 
-        this.context.clearRect(0, 0, this.canvas.width, this.canvas.height);
+            this.context.clearRect(0, 0, this.canvas.width, this.canvas.height);
+        } catch (error) {
+            Sentry.captureException(error);
+            console.error("Error captured in Sentry:", error);
+        }
```
[the remark this answers is not in the document]
Comment (salmoneatenbybear) [reply simulator/src/simulationArea.js:125]: I’m opting not to add error handling to dragStart, panStart, setupEnvironment, and setupElementLists because they’re low-level functions with minimal risk of unexpected inputs. Adding try-catch blocks here could impact performance or mask setup issues. I’ll revisit if needed based on future findings.

for the rest i have implemented the suggestions you gave

File: simulator/src/simulationArea.js [hunk 14]
```diff
@@ -71,38 +72,59 @@
...
+            this.canvas.width = width;
+            this.canvas.height = height;
+            this.simulationQueue = new EventQueue(10000);
+            this.context = this.canvas.getContext('2d');
+            simulationArea.changeClockTime(simulationArea.timePeriod);
+            this.mouseDown = false;
+        } catch (error) {
+            Sentry.captureException(error);
+            console.error("Error captured in Sentry:", error);
+        }
```
[the remark this answers is not in the document]
Comment (salmoneatenbybear) [reply simulator/src/simulationArea.js:97]: done

File: simulator/src/app.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (salmoneatenbybear) [reply simulator/src/app.js:10]: I’m holding off on implementing the suggested environment variable setup (SENTRY_DSN and NODE_ENV) for now, as it requires adjustments in the deployment environment that I’m not ready to configure yet. I plan to revisit this setup when I have a better understanding of our deployment process and the access required to set environment variables securely. For now, I'll continue using the current configuration to maintain progress without interruptions.

File: simulator/src/app.js [hunk 15]
```diff
@@ -1,6 +1,15 @@
+import * as Sentry from "@sentry/browser";
 import { setup } from './setup';
 import Array from './arrayHelpers';
```
[the remark this answers is not in the document]
Comment (salmoneatenbybear) [reply simulator/src/app.js:3]: I’d prefer not to change the import aliasing for Array at this time, as this is how the code was structured before I started working on it. Altering it now might introduce unexpected issues, especially if there are other dependencies or references relying on this specific naming.

File: app/views/simulator/edit.html.erb [hunk 16]
```diff
@@ -1,5 +1,8 @@
 <% content_for(:html_title) { "CircuitVerse - Digital Simulator" } %>
 
+<%# This needs to be replaced with the one from official account one. %>
+<script src="https://js.sentry-cdn.com/f8e920e25d22ae3b791f016f0cb5a2e5.min.js" crossorigin="anonymous"></script>
```
[the remark this answers is not in the document]
Comment (salmoneatenbybear) [reply app/views/simulator/edit.html.erb:4]: I'm holding off on implementing this environment-specific setup as per my previous message regarding SENTRY_DSN configuration adjustments. I’ll revisit it when I'm ready to make changes in the deployment environment.

File: simulator/src/testbench.js [hunk 17]
```diff
@@ -860,29 +1024,33 @@ function triggerReset(reset, scope) {
...
+            const bw = `<td>${escapeHtml(io.bitWidth.toString())}</td>`;
+            $('#tb-manual-table-labels').append(label);
+            $('#tb-manual-table-bitwidths').append(bw);
+        });
 
-    setUICurrentCase(testbenchData);
+        setUICurrentCase(testbenchData);
+    } catch (error) {
+        console.error("Error in setUITableHeaders:", error);
+    }
```
[the remark this answers is not in the document]
Comment (salmoneatenbybear) [reply simulator/src/testbench.js:1053]: I'll hold off on adding user-facing error messages in these UI functions for now. Since we're already capturing these errors in Sentry, we can monitor and address them as needed without potentially confusing the user with technical error messages. Logging to the console combined with Sentry should provide sufficient insight for debugging. If we expand user-friendly error handling across the UI in the future, we can revisit this approach.

File: simulator/src/testbench.js [hunk 18]
```diff
@@ -811,34 +962,43 @@
...
-    scope.clockTick();
-    play(scope);
+    try {
+        scope.clockTick();
+        play(scope);
+        scope.clockTick();
+        play(scope);
+    } catch (error) {
+        console.error("Error in tickClock:", error);
+    }
```
[the remark this answers is not in the document]
Comment (salmoneatenbybear) [reply simulator/src/testbench.js:1001]: done

**Comment by salmoneatenbybear** [comment]
implemented some of the changes the bots said, looks like there are many more, will make them soon...

File: simulator/src/app.js [hunk 19]
```diff
@@ -1,6 +1,10 @@
...
 
+// Initialize Sentry for error tracking
+Sentry.init({dsn: "https://a9660d5d8964d5b348b07b77691d61a7@o4507251968376832.ingest.de.sentry.io/4507251969949776",maxBreadcrumbs: 50,debug: true});
```
[the remark this answers is not in the document]
Comment (salmoneatenbybear) [reply simulator/src/app.js:6]: done

[the remark this answers is not in the document]
Comment (salmoneatenbybear) [reply simulator/src/app.js:6]: done

[the remark this answers is not in the document]
Comment (salmoneatenbybear) [reply simulator/src/app.js:6]: done

[the remark this answers is not in the document]
Comment (salmoneatenbybear) [reply simulator/src/app.js:6]: done

File: simulator/src/canvasApi.js [hunk 20]
```diff
@@ -7,148 +7,135 @@ import { colors } from './themer/themer';
...
+        if (!embed && !lightMode) {
+            findDimensions(globalScope);
+            miniMapArea.setup();
+            $('#miniMap').show();
+            updatelastMinimapShown();
+            setTimeout(removeMiniMap, 2000);
+        }
+    } catch (error) {
+        Sentry.captureException(error);
+        console.error("Error captured in Sentry:", error);
```
[the remark this answers is not in the document]
Comment (salmoneatenbybear) [reply simulator/src/canvasApi.js:85]: intentional

File: simulator/src/canvasApi.js [hunk 21]
```diff
@@ -7,148 +7,135 @@
...
+            }
+            for (var j = 0; j < canvasHeight; j += scale) {
+                ctx.moveTo(0, Math.round(j + correction) - correction);
+                ctx.lineTo(canvasWidth, Math.round(j + correction) - correction);
+            }
+            ctx.stroke();
         }
-        ctx.stroke();
+    } catch (error) {
+        Sentry.captureException(error);
```
[the remark this answers is not in the document]
Comment (salmoneatenbybear) [reply simulator/src/canvasApi.js:136]: intentional

File: simulator/src/canvasApi.js [hunk 22]
```diff
@@ -157,121 +144,199 @@
...
-    y1 *= globalScope.scale;
-    x2 *= globalScope.scale;
-    y2 *= globalScope.scale;
-    x3 *= globalScope.scale;
-    y3 *= globalScope.scale;
-    xx *= globalScope.scale;
-    yy *= globalScope.scale;
-    var ctx = simulationArea.context;
-    ctx.bezierCurveTo(Math.round(xx + ox + x1), Math.round(yy + oy + y1), Math.round(xx + ox + x2), Math.round(yy + oy + y2), Math.round(xx + ox + x3), Math.round(yy + oy + y3));
+    try {
```
[the remark this answers is not in the document]
Comment (salmoneatenbybear) [reply simulator/src/canvasApi.js:147]: done

[the remark this answers is not in the document]
Comment (salmoneatenbybear) [reply simulator/src/canvasApi.js:147]: done

File: simulator/src/canvasApi.js [hunk 23]
```diff
@@ -157,121 +144,199 @@
...
+        [x1, y1] = rotate(x1, y1, dir);
```
[the remark this answers is not in the document]
Comment (salmoneatenbybear) [reply simulator/src/canvasApi.js:148]: done

[the remark this answers is not in the document]
Comment (salmoneatenbybear) [reply simulator/src/canvasApi.js:148]: done

File: simulator/src/simulationArea.js [hunk 24]
```diff
@@ -71,38 +72,61 @@ const simulationArea = {
...
-            simulationArea.clickCount = 0;
-        }, 600);
+        try {
+            clickTimer = setTimeout(() => {
+                simulationArea.clickCount = 0;
+            }, 600);
+        } catch (error) {
+            Sentry.captureException(error);
+            console.error("Error captured in Sentry:", error);
+        }
```
[the remark this answers is not in the document]
Comment (salmoneatenbybear) [reply simulator/src/simulationArea.js:82]: done

File: simulator/src/simulationArea.js [hunk 25]
```diff
@@ -71,38 +72,61 @@
...
-        simulationArea.ClockInterval = setInterval(clockTick, t);
+            clearInterval(simulationArea.ClockInterval);
+            // eslint-disable-next-line no-param-reassign
+            t = t || prompt('Enter Time Period:');
+            simulationArea.timePeriod = t;
+            simulationArea.ClockInterval = setInterval(clockTick, t);
+        } catch (error) {
+            Sentry.captureException(error);
+            console.error("Error captured in Sentry:", error);
+        }
```
[the remark this answers is not in the document]
Comment (salmoneatenbybear) [reply simulator/src/simulationArea.js:115]: done

File: simulator/src/app.js (hunk 19 above)
[the remark this answers is not in the document]
Comment (salmoneatenbybear) [reply simulator/src/app.js:6]: Integrated Sentry error tracking with enhanced configuration for error filtering and data sanitization. Omitted BrowserTracing integration as it was causing conflicts that disrupted simulator functionality. The setup now includes maxBreadcrumbs, debug mode, and beforeSend for selective data filtering.

File: simulator/src/circuit.js [hunk 26]
```diff
@@ -53,93 +54,134 @@
...
+            showProperties(simulationArea.lastSelected);
+            updateTestbenchUI();
+            plotArea.reset();
+        }
+        updateCanvasSet(true);
+        scheduleUpdate();
+        updateRestrictedElementsList();
+    } catch (error) {
+        Sentry.captureException(error);
+        console.error("Error captured in Sentry:", error);
```
[the remark this answers is not in the document]
Comment (salmoneatenbybear) [reply simulator/src/circuit.js:89]: these are just to point the dev to sentry

File: simulator/src/circuit.js [hunk 27]
```diff
@@ -53,93 +54,134 @@
...
+                console.error("Error during DOM manipulation or scope list update:", domError);
+            }
+        } else {
+            showMessage('Circuit was not closed');
+        }
+    } catch (error) {
+        // Capture any unexpected errors in the overall function
+        Sentry.captureException(error);
+        console.error("Unexpected error in deleteCurrentCircuit:", error);
+    }
```
[the remark this answers is not in the document]
Comment (salmoneatenbybear) [reply simulator/src/circuit.js:166]: since this function is central to managing circuits and interacts with several interdependent components, I want to avoid any modifications that might introduce instability or require extensive testing. The current structure, while slightly complex, has been stable and ensures all operations are handled within a single, cohesive function.

File: simulator/src/circuit.js [hunk 28]
```diff
@@ -212,10 +241,15 @@
...
 export function changeCircuitName(name, id = globalScope.id) {
-    name = name || 'Untitled';
-    name = escapeHtml(stripTags(name));
-    $(`#${id} .circuitName`).html(`${truncateString(name, 18)}`);
-    scopeList[id].name = name;
+    try {
+        name = name || 'Untitled';
+        name = escapeHtml(stripTags(name));
+        $(`#${id} .circuitName`).html(`${truncateString(name, 18)}`);
+        scopeList[id].name = name;
```
[the remark this answers is not in the document]
Comment (salmoneatenbybear) [reply simulator/src/circuit.js:248]: done

File: simulator/src/testbench.js [hunk 29]
```diff
@@ -451,18 +542,28 @@ export function runAll(data, scope = globalScope) {
...
+        return result;
+    } catch (error) {
+        // Log the error to Sentry and console for tracking
+        Sentry.captureException(error);
+        console.error("Error in runSingleTest:", error);
+        return null; // Return null or an appropriate default value in case of error
     }
-
-    return result;
 }
```
[the remark this answers is not in the document]
Comment (salmoneatenbybear) [reply simulator/src/testbench.js:564]: The critical error paths are already captured in key functions, and adding Sentry in every function could lead to duplicated logs and noise, making error tracking less efficient. Current coverage is sufficient for tracking major issues.

File: simulator/src/testbench.js [hunk 30]
```diff
@@ -948,64 +1146,77 @@
...
+                    }
+                } catch (error) {
+                    console.error("Error in popup close check:", error);
+                    clearInterval(checkPopUp);
+                }
+            }, 1000);
+        }
+    } catch (error) {
+        console.error("Error in openCreator:", error);
     }
```
[the remark this answers is not in the document]
Comment (salmoneatenbybear) [reply simulator/src/testbench.js:1221]: Introducing an additional cleanup function adds complexity without significantly improving resource management, given the controlled usage of event listeners and intervals in this context.

File: app/views/simulator/edit.html.erb [hunk 31]
```diff
@@ -152,12 +155,11 @@
...
-          <button class="zoom-slider-decrement" id="decrement">-</button>
-          <input type="range" class="form-range" min="0" max="45" step="1" id="customRange1">
-          <span id='slider_value'></span>
-          <button class="zoom-slider-increment" id="increment">+</button>
-        </div>
+        <div class="zoom-slider d-flex align-items-center">
+  <button class="zoom-slider-decrement btn btn-outline-secondary" id="decrement" aria-label="Zoom Out">-</button>
+  <input type="range" class="form-range mx-2" min="0" max="45" step="1" id="customRange1" aria-label="Zoom Level">
+  <button class="zoom-slider-increment btn btn-outline-secondary" id="increment" aria-label="Zoom In">+</button>
+</div>
```
Comment (tachyons) [inline app/views/simulator/edit.html.erb:162]: This is unrelated, create separate Pr for this

Comment (salmoneatenbybear) [reply app/views/simulator/edit.html.erb:162]: Ok noted

File: app/views/simulator/edit.html.erb (hunk 16 above)
Comment (tachyons) [inline app/views/simulator/edit.html.erb:4]: We already have sentry from npm, why do we need CDN version as well ?

File: simulator/src/app.js [hunk 32]
```diff
@@ -1,217 +1,235 @@
...
+                net: 'clk',
+                order: 0,
+                bits: 1,
+            },
+            dev1: {
+                type: 'Input',
+                net: 'addr',
+                order: 1,
+                bits: 4,
+            },
```
Comment (tachyons) [inline simulator/src/app.js:37]: Do not make unrelated changes

File: simulator/src/canvasApi.js [hunk 33]
```diff
@@ -7,148 +7,135 @@ import { colors } from './themer/themer';
...
-                    if (obj.y - obj.upDimensionY < simulationArea.minHeight) { simulationArea.minHeight = obj.y - obj.upDimensionY; }
-                    if (obj.y + obj.downDimensionY > simulationArea.maxHeight) { simulationArea.maxHeight = obj.y + obj.downDimensionY; }
-                    if (obj.x - obj.leftDimensionX < simulationArea.minWidth) { simulationArea.minWidth = obj.x - obj.leftDimensionX; }
-                    if (obj.x + obj.rightDimensionX > simulationArea.maxWidth) { simulationArea.maxWidth = obj.x + obj.rightDimensionX; }
-                } else {
-                    if (obj.absY() < simulationArea.minHeight) { simulationArea.minHeight = obj.absY(); }
-                    if (obj.absY() > simulationArea.maxHeight) { simulationArea.maxHeight = obj.absY(); }
-                    if (obj.absX() < simulationArea.minWidth) { simulationArea.minWidth = obj.absX(); }
-                    if (obj.absX() > simulationArea.maxWidth) { simulationArea.maxWidth = obj.absX(); }
+    try {
```
Comment (tachyons) [inline simulator/src/canvasApi.js:10]: Do not wrap it like this, revert this change
